# Data Cleaning: Creating New Dataset


We have taken our data from the Annual Social and Economic Supplements from the United States' Census Bureau Current Population Survey 2021.

Importing of essential library for data manipulation

In [1]:
import numpy as np
import pandas as pd

In [43]:
dataset = pd.read_csv('ASEC_individual.csv')
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 163543 entries, 0 to 163542
Columns: 830 entries, PERIDNUM to YYYYMM
dtypes: float64(2), int64(827), object(1)
memory usage: 1.0+ GB


There are currently 830 columns and 163543 rows.
By reviewing the data dictionary, we understand that the data is split into 10 broad topics:
1. Record Identifiers
2. Weights
3. Demographics
4. Basic CPS Items
5. Work Experience
6. Income
7. Poverty
8. Health Insurance
9. Supplemental Poverty Measure
10. Migration

Given that we want to find out how one's background affect their salary, we will not be interested in the record identifiers, weight and other social issues and trends like poverty, health insurance, supplemental poverty measure and migration. We will only be interested in topics 3-6.

Within these 6 topics, there may also be columns we are not interested in.
For Demographics , it is split into 3 subtopics:
1. Individual Characteristics
2. Allocation Flags

We are only interested in the individual characteristics. From individual characteristics, we can find out:
1. Highest education attainment (A_HGA) -> EDU
2. Marital status (A_MARITL) -> MARITAL
3. Gender (A_SEX) -> SEX
4. Age (AGE1) -> AGE
5. Race (PRDTRACE) -> RACE
6. Veteran status (PEAFEVER) -> VETERAN
7. Possession of professional certificates (PECERT1) -> PRO_CERT
8. Possession of government certificates (PECERT2) -> GOV_CERT
9. Whether the certification is needed for their job (PECERT3) -> CERT_NEED
10. Grooming disability (PEDISDRS) -> GROOM_DIS
11. Hearing difficulty (PEDISEAR) -> HEAR_DIS
12. Visual disability (PEDISEYE) -> VIS_DIS
13. Conditions affecting ability to run errands (PEDISOUT) -> ERR_DIS
14. Difficulty climbing stairs (PEDISPHY) -> CLIMB_DIS
15. Difficulty concentrating (PEDISREM) -> CONCEN_DIS
16. Citizenship group (PRCITSHP) -> CITIZENSHIP

We can then create a list containing the relevant columns names for demographics

In [35]:
demographics = ['A_HGA', 'A_MARITL', 'A_SEX', 'AGE1', 'PRDTRACE', 'PEAFEVER', 'PECERT1', 'PECERT2', 'PECERT3', 'PEDISDRS', 'PEDISEAR', 'PEDISEYE', 'PEDISOUT', 'PEDISPHY', 'PEDISREM', 'PRCITSHP']

#new column names for demographics
demographics_new = ['EDU', 'MARITAL', 'SEX', 'AGE', 'RACE', 'VETERAN', 'PRO_CERT', 'GOV_CERT', 'CERT_NEED', 'GROOM_DIS', 'HEAR_DIS', 'VIS_DIS', 'ERR_DIS', 'CLIMB_DIS', 'CONCEN_DIS', 'CITIZENSHIP']

For Basic CPS Items, we can find out:
1. Hours worked per week (A_HRS) -> HrsWeek
2. Industry (A_MJIND) -> Industry
3. Occupation (PEIOOCC) -> Occupation
4. Class of work; private, government etc (PEIO1COW) -> WorkClass

For Work Experience, we can find out:
1. Time spent looking for job (WELKNW) -> JobSearchTime

For Income, we can find out:
1. Total wage and salary earnings (WSAL_VAL) -> Salary

We can then create a list containing the relevant columns names for Basic CPS, experience and income

In [36]:
basic_cps = ['A_HRS1', 'A_MJIND', 'PEIOOCC', 'PEIO1COW']
experience = ['WELKNW']
income = ['WSAL_VAL']

#new column names for basic_cps, experience and income
basic_cps_new = ['HrsWeek', 'Industry', 'Occupation', 'WorkClass']
experience_new = ['JobSearchTime']
income_new = ['Salary']

In [38]:
#create new dataset
new_data = pd.DataFrame()
new_cols = [demographics_new, basic_cps_new, experience_new, income_new]

n = 0
for lst in [demographics, basic_cps, experience, income]:
    new_lst = new_cols[n]
    n += 1
    for i in range(len(lst)):
        new_data[new_lst[i]] = dataset[lst[i]]
new_data

EDU  MARITAL  SEX  AGE  RACE  VETERAN  PRO_CERT  GOV_CERT  CERT_NEED  \
0        39        1    2   12     1        2         2        -1         -1   
1        39        1    1   12     1        2         2        -1         -1   
2        39        4    2   17     1        2         2        -1         -1   
3        43        1    2   15     1        2         2        -1         -1   
4        39        1    1   15     1        2         1         1          1   
...     ...      ...  ...  ...   ...      ...       ...       ...        ...   
163538   39        7    1   15     4        2         1         1         -1   
163539   39        1    1   16     4        2         2        -1         -1   
163540   39        1    2   15     4        2         2        -1         -1   
163541   40        1    1   12     1        2         2        -1         -1   
163542   44        1    1   11     1        2         2        -1         -1   

        GROOM_DIS  ...  ERR_DIS  CLIMB_DIS  CONCEN_DIS  CITIZENSHIP  HrsWeek  \
0               1  ...        1          1           1            1        0   
1               2  ...        2          2           2            1        0   
2               2  ...        2          2           2            1        0   
3               2  ...        2          2           2            1       32   
4               2  ...        2          2           2            1       40   
...           ...  ...      ...        ...         ...          ...      ...   
163538          2  ...        2          1           2            1        0   
163539          2  ...        2          2           2            4        0   
163540          2  ...        2          2           2            4       40   
163541          2  ...        2          2           2            1        0   
163542          2  ...        2          2           2            1       40   

        Industry  Occupation  WorkClass  JobSearchTime  Salary  
0              0          -1          0              1       0  
1              3        6305          4              7   10000  
2              0          -1          0              1       0  
3             10        2002          5              7   43000  
4              6        9130          4              7   33000  
...          ...         ...        ...            ...     ...  
163538         0          -1          0              7       0  
163539         0          -1          0              7       0  
163540        11        4230          4              7   35000  
163541         0          -1          0              1       0  
163542        10        3322          4              7   43000  

[163543 rows x 22 columns]

In [40]:
new_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 163543 entries, 0 to 163542
Data columns (total 22 columns):
 #   Column         Non-Null Count   Dtype
---  ------         --------------   -----
 0   EDU            163543 non-null  int64
 1   MARITAL        163543 non-null  int64
 2   SEX            163543 non-null  int64
 3   AGE            163543 non-null  int64
 4   RACE           163543 non-null  int64
 5   VETERAN        163543 non-null  int64
 6   PRO_CERT       163543 non-null  int64
 7   GOV_CERT       163543 non-null  int64
 8   CERT_NEED      163543 non-null  int64
 9   GROOM_DIS      163543 non-null  int64
 10  HEAR_DIS       163543 non-null  int64
 11  VIS_DIS        163543 non-null  int64
 12  ERR_DIS        163543 non-null  int64
 13  CLIMB_DIS      163543 non-null  int64
 14  CONCEN_DIS     163543 non-null  int64
 15  CITIZENSHIP    163543 non-null  int64
 16  HrsWeek        163543 non-null  int64
 17  Industry       163543 non-null  int64
 18  Occupation     163543 no

We will further reduce the dataset size by only including those who are working (ie salary > 0)

In [44]:
new_data = new_data[new_data['Salary'] > 0]

Let us see the new dataset information

In [45]:
new_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 76869 entries, 1 to 163542
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   EDU            76869 non-null  int64
 1   MARITAL        76869 non-null  int64
 2   SEX            76869 non-null  int64
 3   AGE            76869 non-null  int64
 4   RACE           76869 non-null  int64
 5   VETERAN        76869 non-null  int64
 6   PRO_CERT       76869 non-null  int64
 7   GOV_CERT       76869 non-null  int64
 8   CERT_NEED      76869 non-null  int64
 9   GROOM_DIS      76869 non-null  int64
 10  HEAR_DIS       76869 non-null  int64
 11  VIS_DIS        76869 non-null  int64
 12  ERR_DIS        76869 non-null  int64
 13  CLIMB_DIS      76869 non-null  int64
 14  CONCEN_DIS     76869 non-null  int64
 15  CITIZENSHIP    76869 non-null  int64
 16  HrsWeek        76869 non-null  int64
 17  Industry       76869 non-null  int64
 18  Occupation     76869 non-null  int64
 19  Wor

Therefore, there are 76869 working adults' data for us to work with.

We will export this dataframe into a seperate csv file for further preprocessing.

In [47]:
new_data.to_csv('reduced_dataset_for_preprocessing.csv', index=False)